Add initial support for JLC PCB assembly

yaqwsx opened the pull request:
Both BOM and POS files are pretty easy to implement, therefore, it makes sense not to wait for Python API and implement the logic in KiKit. To get the BOM file, we need to parse the schematic file, which is currently pretty easy to do (and when we switch to KiCAD 6, we use Eeschema Python API to generate it)

I am looking for testers before I merge this functionality into master. If you are interested in testing, install development version of KiKit by:
```
pip3 install git+https://github.com/yaqwsx/KiKit@fabrication
```

The usage should be pretty straightforward (but currently lacks documentation). Just run `kikit fab jlcpcb --help`. KiKit expects there is a field "LSCS" with LSCS number in for the components in schematics.

Please, report both working and not working cases.

mayl commented:
This looks really promising, thanks!  I've downloaded it and tried it against the project I panelized with kikit before.  Unfortunately, I haven't gotten it to work yet on my design, but I still have have some initial thoughts/questions:

1) The first issue is that I get an error: 
```kikit.eeshema.EeschemaException: Unexpected line: 'AR Path="/5F6CB909" Ref="R?"  Part="1"'```
This seems to be related to [ar path](https://forum.kicad.info/t/purpose-of-unit-path-entries-in-schematic/8254/3) which is related to my use of hierarchical sheets, or possibly a corruption in my eeschema file.  I have done no further troubleshooting yet.  Have you tried the `kikit fab` command with a hierarchical design?  I'm not sure if this is an issue with how my project is set up, or an edge case that the `eeschema` python api is not handling correctly...

2) I think a low impact, but useful feature would be the ability to optionally override the default attribute name of "LSCS" to something else on the command line.  This way, people who have already implemented their library with a different attribute name can easily use Kikit without needing to tediously rework every component in their library.  While we're at it, it would be nice to specify a _list_ of attribute names, so that parts can be used from multiple sources which may have inconsistent conventions. 

2) Minor nit, but manually listing out schematic files is a bit annoying.  Looking at my project, it seems like each schematic contains a relative path to any sub-sheets.  Perhaps an item for further improvement could be to have Kikit automatically walk down the list of schematics and collect the files (or maybe the upcoming Kicad 6 api will just provide it to us).

I'll try and poke a bit more at the "AR PATH" thing and see what I can figure out.  I think this looks like a really cool feature for Kikit, so let me know if there's anything else I can do to help!

yaqwsx commented:
Thank you for your testing - you are right, I haven't tested it with hierarchical sheets. I agree that KiKit has to traverse them automatically. I'll look at it.

iromero91 commented:
Other projects like ibom use an exported netlist instead of trying to analyze the schematic directly. Maybe that could be a good starting point. It is a little bit more kludgy to use as the user has to manually export the netlist. but it will be more robust against corner cases. and hopefully with kicad6 the kludge will be gone for good.

yaqwsx commented:
> It is a little bit more kludgy to use as the user has to manually export the netlist.

And this is a showstopper. If I wouldn't mind the manual export of the netlist, I wouldn't write my own tool and use the tools out there. KiKit aims at **full** automation of the post-processing/exporting process.

mayl commented:
> Other projects like ibom use an exported netlist instead of trying to analyze the schematic directly. Maybe that could be a good starting point. It is a little bit more kludgy to use as the user has to manually export the netlist. but it will be more robust against corner cases. and hopefully with kicad6 the kludge will be gone for good.

I'm not sure how @yaqwsx feels about it, but one thing I like about KiKit is that it is scriptable/accessable from the command line, which makes it much more amenable to automation.  What I'm looking forward to is having a `makefile` for my Kicad projects, that goes from a design, to a panelized design package to submit for quotes in a single command. Having to manually export a BOM from within eeschema would break that possibility and be a disappointing solution IMO.

iromero91 commented:
To be honest I really do like the automation aspect. But this is one thing that really needs some support from the KiCAD side to do right. If you don't mind chasing bugs from corner cases from parsing the schematic files by all means do so, it's more powerful and will allow us to have full functionality right now.

yaqwsx commented:
@mayl I implemented the hierarchical sheets. Could you test it? I don't have any hierarchical boards specifically designed for JLC, so I tested it only on a simple dummy project.

mayl commented:
With this latest commit `kikit fab` processes my design without error  :+1: 

I haven't checked if any of the files are correct, but it's getting late here and that'll have to wait until tomorrow.  I'll take a closer look at them and try putting them through the JLC online quote system and reply back.

mayl commented on kikit/fab/jlcpcb.py:
```diff
@@ -0,0 +1,90 @@
+        writer = csv.writer(csvfile)
+        writer.writerow(["Comment", "Designator", "Footprint", "LCSC"])
+        for cType, references in bomData.items():
+            value, footprint, lscs = cType
+            writer.writerow([value, ",".join(references), footprint, lscs])
```
```suggestion
            value, footprint, lcsc = cType
            writer.writerow([value, ",".join(references), footprint, lcsc])
```

mayl replied:
I'm glad to see I'm not the only one who finds this acronym confusing, but it's technically "lcsc" :smile:

yaqwsx replied:
Thanks. It is extremely confusing...

mkousal commented:
I tested it and it works fine for the first try, but one thing is wrong. Part number isn't LSCS, but LCSC. I don't know if KiKit search for this column by default or if it must be specified every time in CLI

mayl commented on kikit/eeshema.py:
```diff
@@ -0,0 +1,161 @@
+        line = readEeschemaLine(file)
+        if line == "$EndSheet":
+            return sheet
+        if line.startswith("F1"):
```
```suggestion
        if line.startswith("F1 "):
```
Match the start of line exactly.  My design has "F10, F11, F12" and so on and this matches those as well.

mayl commented:
I've been playing with this a little bit today, but had a couple issues still.  Firstly, see my suggested fix to find the F1 line correctly in the parser.  Some of my subsheets have more than 10 interconnects, and this match was too permissive.

Secondly, there still seems to be an issue with the `AR` that I haven't quite figured out that causes duplicate reference designators to be extracted.  It seems like components which have a hierarchical net associated with them get duplicate "unassigned" reference designators.  Here are a few examples:
```
$Comp
L Device:R_US R49
U 1 1 5F9AF435
P 4350 2300
AR Path="/5F9AF435" Ref="R49"  Part="1" 
AR Path="/5F63AB1F/5F9AF435" Ref="R?"  Part="1" 
F 0 "R49" V 4555 2300 50  0000 C CNN
F 1 "10k" V 4464 2300 50  0000 C CNN
F 2 "Resistor_SMD:R_0402_1005Metric" V 4390 2290 50  0001 C CNN
F 3 "~" H 4350 2300 50  0001 C CNN
F 4 "C25744" H 4350 2300 50  0001 C CNN "LCSC_PN"
F 5 "0402WGF1002TCE" H 4350 2300 50  0001 C CNN "MPN"
F 6 "Y" H 4350 2300 50  0001 C CNN "basic?"
F 7 "0.0009" H 4350 2300 50  0001 C CNN "price"
F 8 "C25744" H 4350 2300 50  0001 C CNN "Field4"
F 9 "0402WGF1002TCE" H 4350 2300 50  0001 C CNN "Field5"
F 10 "C25744" H 4350 2300 50  0001 C CNN "LCSC Part#"
	1    4350 2300
	-1   0    0    1   
$EndComp
```
```
$Comp
L Device:C C?
U 1 1 5F0DD2F8
P 1550 1350
AR Path="/5F0DD2F8" Ref="C?"  Part="1" 
AR Path="/5F0B66EA/5F0DD2F8" Ref="C?"  Part="1" 
AR Path="/5F63AB1F/5F780AD8/5F0DD2F8" Ref="C20"  Part="1" 
F 0 "C20" H 1665 1441 50  0000 L CNN
F 1 "10uF" H 1665 1350 50  0000 L CNN
F 2 "Capacitor_SMD:C_0402_1005Metric" H 1588 1200 50  0001 C CNN
F 3 "~" H 1550 1350 50  0001 C CNN
F 4 "CL05A106MQ5NUNC" H 1550 1350 50  0001 C CNN "MPN"
F 5 "C15525" H 1550 1350 50  0001 C CNN "LCSC_PN"
F 6 "6.3V" H 1665 1259 50  0000 L CNN "Voltage"
F 7 "Y" H 1550 1350 50  0001 C CNN "basic?"
F 8 "0.0144" H 1550 1350 50  0001 C CNN "price"
F 9 "C15525" H 1550 1350 50  0001 C CNN "Field4"
F 10 "CL05A106MQ5NUNC" H 1550 1350 50  0001 C CNN "Field5"
F 11 "C15525" H 1550 1350 50  0001 C CNN "LCSC Part#"
	1    1550 1350
	1    0    0    -1  
$EndComp
```
This leads to my BOM and POS files to have multiple "fake" reference designators like "R?" and "C?".  I can _sort of_ get around this by only adding reference designators that don't end in question mark, but this seems like a weak heuristic.

yaqwsx commented:
@mayl Could you provide me your schematics? Currently, I've looked at all of my projects with hierarchical sheets and none of them have the "dangling references" as you show. I would like to find out the root cause of them by browsing the schematics in KiCAD.

mayl commented:
Unfortunately, I'm unable to share the files from my actual project, but attached is a simple case which demonstrates the same issue. This is a snippet from `subsheet.sch`
```
$Comp
L Device:C_Small C?
U 1 1 5F171610
P 3050 1700
AR Path="/5F171610" Ref="C?"  Part="1" 
AR Path="/5F16D08F/5F171610" Ref="C4"  Part="1" 
F 0 "C4" H 3142 1746 50  0000 L CNN
F 1 "C_Small" H 3142 1655 50  0000 L CNN
F 2 "" H 3050 1700 50  0001 C CNN
F 3 "~" H 3050 1700 50  0001 C CNN
	1    3050 1700
	-1   0    0    -1  
$EndComp
```

[hierarchicalTest.zip](https://github.com/yaqwsx/KiKit/files/4951972/hierarchicalTest.zip)

yaqwsx commented:
I understand the problem now. Did you first created `subsheet.sch` and then turn it into a hierarchical subsheet or opened the subsheet directly from Eeschema? Looking at the format, the dangling references (containing references with `?`) come from the time, when `subsheet.sch` was considered as a top-level/standalone sheet. This is why I wasn't unable to reproduce your issue. 

The solution is easy - we just have to track the path to the subsheet. I will implement it soon.

mayl commented:
I don't really think I did either?  I produced the top level sheet, then placed a hierarchical sheet label with a new sheet name, alt-double clicked into it to draw the sub-schematic and then saved all.  In either case, I'm glad you understand the fix.

yaqwsx commented:
I am not sure why the dangling references appear in your case and not mine. I use KiCAD 5.1.6 - maybe there is a different behavior among versions? The dangling references don't change the semantics of the schematic if you read them properly (I was reading them improperly to make it simpler) so the different behavior might be plausible. I pushed the fix. It seems to work on your minimal counterexample.

Also, I know there were different opinions on rotation and origin compensation in #34 - some of you suggested hardcoded corrections for built-in footprints, some of you claimed that the users should build their own libraries.

What about a compromise? I like the way https://github.com/matthewlai/JLCKicadTools/blob/master/jlc_kicad_tools/cpl_rotations_db.csv handles it. There is a CSV file with regex patterns for footprint names and compensation value (both rotation and origin). This file can be supplied from command line. We could also provide a default compensation file for KiCAD libraries with KiKit.

What is your opinion on this?

mayl commented:
I don't know why either, I'm using KiCAD 5.1.6 as well (on linux if that matters?).  

Your latest commit seems to work for my project, no more "dangling references" in the BOM.  However, I've noticed another issue.  For parts with multiple "Units" (such as a 4 channel comparator), the reference designator is duplicated in the BOM.  I updated my simple test to demonstrate this as well: [hierarchicalTest.zip](https://github.com/yaqwsx/KiKit/files/4952392/hierarchicalTest.zip)

With respect to the rotation issue, I still prefer specifying a "JLC_Rotation" or similar attribute in the parts themselves.  I am of the opinion that the "db"/regex approach is fragile and less desirable for a couple reasons.

1) You now have an additional file which determines the assembly package.  If this is a "default" file which ships with KiKit, and KiKit updates it, old projects could be broken.

2) Writing and supplying your own offset file is another thing users will need to understand and do.  They already have to add attributes to the parts in their designs to insert LCSC part numbers, so including any assembly offsets there as well would be more consistent I think.

3) The regex works if you are using purely the KiCAD default libraries, but may not work if you need to use parts from multiple sources and two or more parts match the same regex, but need different offsets.

This isn't to say I understand the appeal of the regex approach, for the cases it works for it's easier, but I think in the long run it is less flexible and more fragile than the part attribute approach.

yaqwsx commented:
@mayl I added code handling multi-units components.

I like your proposal with extra fields. It's better in most of the aspects. To handle existing designs, we could add a tool to KiKit to automatically annotate the schematic based on regex rules.

mayl commented:
Awesome, the multi-unit handling works for my larger design as well!

samkristoff commented:
Hey,

I played around with this PR on a simple board and it worked as expected when generating output for a single board.

I'd like to take my single board, panelize it and generate the gerbers for the panel, bom for a single board and position data for the single board.

**What I tried:**
Panelize the board using `kikit panelize ...` 
Point `kikit fab jlcpcb` at my schematic and the panelized PCB
**Result:**
Gerbers for panel
BOM for 1 board
pos.csv for entire panel (with duplicate reference designators)

I'm not sure what the 'right' or 'best' behavior is in general, but for me it would be ideal for `kikit fab` to output pos.csv containing coordinates for a single board.

A work around is to panelize, generate the gerbers using `kikit export gerber ...` (using the panel) and then use `kikit fab` to generate the pos.csv file using the single board PCB file.  In this case it would be useful to be able to suppress gerber generate in `kikit fab` since it will output gerbers for a single board

mayl commented:
Interesting, what is your use case for panelizing a design generating a pick and place file for just one of the boards?  In any event, can't you achieve this already by calling `kikit fab` on your single board?

samkristoff commented:
I just ordered a Neoden 4 pick and place machine and based on what I've read so far it seems it expects component position data for a single board and number of boards (x and y) in the panel.  I think it will populate one board to completion before moving on to the next board and will calibrate using each board's feducials before populating.

I'm not sure if the Neoden 4 will object to the same component reference designator showing up multiple times in the pos.csv file.  If not it might be possible to have the Neoden treat the panel as a single board.  I should know more in a few days when the system arrives.

yaqwsx commented:
@samkristoff I do not plan to add such feature for two reasons:
- first, I would like to keep CLI simple and you can easily obtain your files by calling KiKit twice - just like mayl says
- second, I decided to keep the fab command vendor-specific - i.e., `kikit fab jlcpcb` should output files that will be accepted by JLC PCB. If you are interested, feel free to submit a PR for `kikit fab pnp --machine neoden4` or similar. I will accept such PR.

I am on a vacation for the next week. Once I return, I plan to implement the component rotation and origin compensation and then merge this PR.

samkristoff commented:
Makes sense.  If I get something working I'll submit a PR.  Thanks for your work on KiKit, it's been very helpful!

yaqwsx commented:
@mayl I implemented the corrections for the assembly. It allows you to translate component origin and its rotation. You specify it in field `JLCPCB_CORRECTION` in format `<X>; <Y>; <orientation`. Please test it. If you find no bugs, I'll write documentation and merge the feature.

mayl commented on kikit/fab/jlcpcb.py:
```diff
@@ -3,30 +3,64 @@
-def moduleX(module, placeOffset):
+def modulePosition(module, placeOffset, compensation):
     pos = module.GetPosition() - placeOffset
+    angle = -radians(module.GetOrientation() / 10.0)
```
Maybe I'm not thinking through this correctly, but if you're making a rotation correction and an offset correction, is it correct to apply the rotation before offsetting the center of rotation like this?  I'm having trouble visualizing precisely how I might use the x,y correction feature, but I *think* it would be to bring an incorrect part's origin into the parts actual center of rotation.  I would probably want to apply any rotation offset after that?

yaqwsx replied:
The use case for XY correction comes from solder jumpers. In some of my designs, I use solder jumpers - a 3 pad component where you can place zero-ohm resistor to two places. The solder jumper has the origin in its center. By using 0402 resistors and by applying the correction I can select which configuration will be populated by the assembly house.

mayl replied:
Hmm, I see.  I don't know what facilities KiCAD makes available, but personally I would want something like this visible from the schematic.  In the past, I've used the solder jumpers which is not a BOM item and can be "programmed" through selecting the right footprint.  Personally, specifying an offset in the `JCLPCB_CORRECTION` (manufacturer specific) to create a connectivity change seems to mix multiple concerns in a rather opaque way that I would tend to avoid, but maybe this is the only way based on your workflow?  In any case, I think it's important to note in the documentation the order the correction is applied (rotate, then translate).

yaqwsx replied:
The problem with having multiple footprints and changing them is that they have different origins and your footprint will move when you change it, therefore, you cannot automate it. I would like to easily produce different configurations of the boards. When I think about it, it would be nice to have an option to specify list of fields with the corrections where the first found is used. Then you can have fields like `JLCPCB_CORRECTION` for all components and for some `JLCPCB_CORRECTION_CFG1` and `JLCPCB_CORRECTION_CFG2`. This way it is transparent what the offset means.

Also one note for not going the footprint way - the connectivity is the matter of assembly more than board layout. With a different manufacturer, you might want to go a different way.

You are right about the documentation. I'll make sure I put some examples there.

mayl replied:
Not trying to convince you of anything, but just to clarify what I meant: changing the footprint to "reprogram" the jumper was in the case of the solder jumpers.  These can have the same origin and rotation, but with a footprint with option A connected or option B connected (in copper).  These are just defaults, and can be reprogrammed by cutting the trace in the jumper and solder bridging the two other contacts.  I see the difficulty in doing this if you want to populate 0 Ohm Resistors, I was just trying to propose an alternate approach.

I also don't think I agree with your approach to connectivity.  To me, all the connectivity should be captured in the schematic, since that is ultimately what the schematic is for.  I think when you change manufacturers in KiKit, you should get the same design just targeted to a different manufacturer.  IMO, which jumper options are selected should be captured in the design somewhere, not in the translation from design to manufacturing package.

Thinking about it a bit more, I think there are other problems with the approach too.  If you have a design with 3 jumpers with the same footprint, I don't see how you can independently change what each is programmed to (JP1=A, JP2=B, and so on)?

I hope this comment didn't read as overly critical, I don't mean to argue against your workflow, especially since I don't have a strong alternative to offer.  However, I am still unconvinced that using the per-manufacturer placement translation in this way is a good idea and wanted to give you my reasons so you can consider them.

yaqwsx replied:
I don't take it critically and I am open to other opinions and views.

IMHO the connectivity is captured in the schematics - it is a jumper after all. I just program it to a default value (possibly based on a text note in schematics). The corrections are per instance of a component, not per footprint, so you can easily do JP1=A, JP2=B.

mayl replied:
I think I see one of the disconnects.  To me, I would save the manufacturer corrections back to my library.  That way in the future, if I use that part it already has an `LCSC part number`, `JLCPCB_CORRECTION` etc. assigned.  This is pretty incompatible with tweaking `JLCPCB_CORRECTION` to reprogram jumpers.  It sounds like your approach is to tweak all the corrections anyway, so what is baked in as the default is less of an issue for you.  Translating text notes to offsets and repeating this for every manufacturer you want to support, for every jumper in every design still seem tedious and error prone to me, but maybe it's the only way.

Anyway, it's easy enough for me to just not use the translation feature so I wouldn't take my opinion as a reason to block merging this :smile:

mayl commented:
I used this to take my design all the way through panelization, bom and pos file generation and did a quote through JLCPCB successfully, so this looks good :+1: !

I'm still a little put off by the x,y translation correction.  Unlike rotation where multiple conventions exist, I have only seen the origin of the part being off the center of rotation as a bug in the library.  Also see my comment about the order of operations.  I don't have any parts which exercise this feature, so whether or not the problem identified is an issue is not certain.  The syntax you chose for this is very clear and easy to use, so typing the extra information in isn't an issue.

A second general thought is that perhaps this field should be overridable with a command line switch as well?

yaqwsx commented:
I added the documentation. Could you, @mayl, look at it before I merge the feature?

mayl commented:
Hi @yaqwsx the documentation reads very clearly to me, I'm excited to see this feature merged!

Until I read your documentation, I hadn't realized that only parts with the `LCSC` field were added to the BOM.  I'm not sure if that is desirable.  For certain generic components like resistors or capacitors, I sometimes prefer to just leave them as a generic component and resolve the exact part number based on cost and available stock when quoting through the JLCPCB website.  A feature for a future time perhaps, but maybe KiKit should print a warning about parts without a BOM entry? I can just see forgetting to assign `LCSC` to something like `R102` in your big design and getting boards back with it missing.  There's not really anything in either KiKit or the quoting process that I think will help you notice that it's absent from the assembly information, so it's really on the designer to double check everything.

In any case, it's great to have the documentation on this so at least anyone who reads the docs should know to flag every part with an `LCSC` field (even if it's a placeholder that you'll refine later in the quote).

yaqwsx commented:
You are right - it would be nice to prevent this kind of mishaps. Since we already have the feature to ignore components, I think it might be useful to have option to treat missing LCSC field as an error. That's handy for CI and automatically generated sites to download manufacturing data. I'll implement it later today.

yaqwsx commented:
Eeh, there was a bug in the documentation -- all the components are included. I fixed it. Nevertheless, I added a warning on the missing field and I implemented option to treat this warning as an error.

So I guess we are ready to for merge!
